### PR TITLE
Add support for boolean flags for arguments

### LINKED ISSUE #2898: [FR] Run Docker with -t argument
**Please fill in this feature request template to ensure a timely and thorough response.**

## Willingness to contribute

- [ ] Yes. I can contribute this feature independently.
- [X] Yes. I would be willing to contribute this feature with guidance from the MLflow community.
- [ ] No. I cannot contribute this feature at this time.

## Proposal Summary

Please enable the -t flag for docker run.

## Motivation
To enable colored output inside a run Docker container I need the `-t` flag. This requires a tty.

Colored output is very helpful to discern status and error messages when training models.

### What component(s), interfaces, languages, and integrations does this feature affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interfaces
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [X] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Languages 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

## Details

The corresponding code is here: https://github.com/mlflow/mlflow/blob/6ef63a7f33ebfe3481ea57cacd94a8a32fc7efd8/mlflow/projects/__init__.py#L528

The cmd needs to (optionally) expand from `docker run --rm` to `docker run --rm -t`

### PR DESCRIPTION
**Pull request by Zethson**
## What changes are proposed in this pull request?

Closes #2898 

Added a special docker argument 'tty', which can be passed as all other Docker arguments via `-A tty=true`.
This enables pseudo TTY support -> color support for example.
I did not want to add a parameter somewhere else, since I think this would get confusing quickly and require more code changes. 

## How is this patch tested?

Tested locally with a container that has colored output. 

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Enabled pseudo TTY support for Docker based projects via special Docker argument `-A tty=true`.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [X] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [X] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by Zethson** [comment]
Requesting review from @dbczumar .

**Comment by Zethson** [comment]
Kindly requesting review from @dbczumar.

File: mlflow/cli.py [hunk 1]
```diff
@@ -53,7 +53,8 @@ def cli():
                    "corresponding entry point as command-line arguments in the form `--name value`")
 @click.option("--docker-args", "-A", metavar="NAME=VALUE", multiple=True,
               help="A `docker run` flag or argument, of the form -A name=value. Where `name` "
-              "will then be propagated as `docker run --name value`.")
+              "will then be propagated as `docker run --name value`. "
+              "Pass -A tty=true to allocate a pseudo-TTY.")
```
Comment (dbczumar) [inline mlflow/cli.py:57]: @Zethson Instead of renaming / aliasing the `-t` argument of `docker run` as `-tty`, can we continue to use `-t`? This way, we don't have to do the special casing in `mlflow.projects.__init__`. This will likely be clearer to users who are familiar with the `-t` flag in Docker.

Comment (Zethson) [reply mlflow/cli.py:57]: Sure. Will not look that descriptive: `-A t=true`, but fine with me if you prefer it.

Comment (dbczumar) [reply mlflow/cli.py:57]: Awesome! Thanks!

File: mlflow/cli.py [hunk 2]
```diff
@@ -53,7 +53,8 @@ def cli():
                    "corresponding entry point as command-line arguments in the form `--name value`")
 @click.option("--docker-args", "-A", metavar="NAME=VALUE", multiple=True,
               help="A `docker run` flag or argument, of the form -A name=value. Where `name` "
-              "will then be propagated as `docker run --name value`.")
+              "will then be propagated as `docker run --name value`. "
+              "Pass -A t=true to allocate a pseudo-TTY.")
```
Comment (dbczumar) [inline mlflow/cli.py:57]: @Zethson Ah, I see the difficulty now. I was hoping to allow users to simply pass the `-t` flag as `--docker-args t`, but it looks like our `NAME=VALUE` metavar structure prevents that from happening. Rather than special casing `t=true` or `tty=true`, perhaps we can revise `--docker-args` to take either a pair of `NAME=VALUE` or a boolean flag of the form `NAME`. That way, users could enable a tty using one of the following:

```
mlflow run --docker-args tty
mlflow run --docker-args t
```

This would automatically provide support for all other boolean flags as well. Does that approach make sense?

Comment (Zethson) [reply mlflow/cli.py:57]: Yes, of course. This is also how I would have done it initially, but I didn't quite know how many things I am allowed to touch :)
Will implement tomorrow.

File: mlflow/cli.py [hunk 3]
```diff
@@ -136,16 +137,23 @@ def run(uri, entry_point, version, param_list, docker_args, experiment_name, exp
...
-        index = s.find("=")
-        if index == -1:
+    for arg in arguments:
+        split = arg.split('=')
+        # Docker arguments such as `t` don't require a value
+        # -> set them to True if specified
+        if len(split) == 1:
+            if argument_type == 'A':
+                name = split[0]
+                value = True
```
Comment (Zethson) [inline mlflow/cli.py:149]: Fine with you if I work with booleans here or do you want me to use 'true'?

Comment (dbczumar) [reply mlflow/cli.py:149]: @Zethson How do we intend to map True (or `true`) to the proper command flag (which is just `-t` rather than `-t True` or `-t true`)?

Comment (dbczumar) [reply mlflow/cli.py:149]: Perhaps the user args to do not necessarily have to be a dictionary, since certain args do not have value elements. We may consider changing the return type of this to a list of `UserArgument` objects with two properties: `name` (required) and `value` (optional).

Comment (Zethson) [reply mlflow/cli.py:149]: Not sure whether I understand the first question.
If only a single parameter is specified without `=value`, it is automatically set to the value True. This only happens when the parameter is passed. All command line interfaces treat options like this. Note that this behavior is only enabled for Docker arguments and not for mlflow -P parameters. 

About the second question: If mlflow were at Python 3.7 we could use a pretty [data class](https://docs.python.org/3/library/dataclasses.html). However, since this is not the case I guess it would not make the code much prettier. 

>  since certain args do not have value elements. 

Well, yes and no. They don't have an associated value when passed by the user, but they automatically get the value True inside mlflow. At runtime every argument has a name AND a value.

I am fine with the current design, but if you have more specific wishes (and reasoning) I am of course happy to adjust the implementation.

Comment (dbczumar) [reply mlflow/cli.py:149]: > If only a single parameter is specified without `=value`, it is automatically set to the value True. 

I think the issue is that this behavior is incompatible with downstream CLI tools like Docker. For example, if `_user_args_to_dict` returns a dictionary with a string -> boolean mapping which is then passed to `_get_docker_command`, the resulting Docker command will contain an argument of the form `--name True`, which is not the correct behavior for boolean flags on the commandline.

`docker run --interactive True ...` is *invalid*
`docker run --interactive ...` is valid.

I think defining `user_args_to_dict` like this is fine, but the downstream consumer of this dictionary (i.e., `_get_docker_command`) needs to handle boolean values in the dictionary. Specifically, the string `True` or `False` should not be passed to the `docker run` command. Let me know if this makes sense.

File: mlflow/projects/__init__.py [hunk 4]
```diff
@@ -534,6 +534,13 @@ def _get_docker_command(image, active_run, docker_args=None, volumes=None, user_
     cmd = [docker_path, "run", "--rm"]
 
     if docker_args:
+        # Allocate a pseudo-TTY if specified
+        if 't' in docker_args:
```
Comment (dbczumar) [inline mlflow/projects/__init__.py:538]: Once the changes described in https://github.com/mlflow/mlflow/pull/3061/files#r451860120 have been made, I don't think we'll need this special casing logic for `t`. We can probably remove this code.

File: mlflow/cli.py (hunk 3 above)
Comment (Zethson) [reply mlflow/cli.py:149]: Yeah, I'm aware of that. Hence, the manual setting of the -t in the command.

The only thing that makes this more complicated is that all the options that may contain boolean values need to be passed BEFORE the rest of the a=b arguments.

Hence, I suggest that I write some code that filters all pairs of name:booleanvalue -> set them first by appending `-name` to the docker run command -> removing them from the dict (or use a prefiltered one) and then append the rest as is business now.
Since the dict is always super small there is no performance penalty anyways.

Does this sound fine to you?

Comment (Zethson) [reply mlflow/cli.py:149]: This would of course also allow for all boolean options (not just -t like now).
Though I fail to see why you would ever pass -i  :)

Comment (dbczumar) [reply mlflow/cli.py:149]: Hmmm...I'm probably missing something, but I don't understand why boolean arguments need to come first. Can you clarify?

Comment (Zethson) [reply mlflow/cli.py:149]: Huh, actually I thought so, but this doesn't seem to be the case.
Well yeah, then things get easier of course.

Comment (Zethson) [reply mlflow/cli.py:149]: I need to sleep, but will get back to you with a new, certainly cleaner looking draft tomorrow.

Comment (dbczumar) [reply mlflow/cli.py:149]: Awesome!

File: mlflow/projects/__init__.py [hunk 5]
```diff
@@ -534,8 +534,13 @@ def _get_docker_command(image, active_run, docker_args=None, volumes=None, user_
     cmd = [docker_path, "run", "--rm"]
 
     if docker_args:
-        for key, value in docker_args.items():
-            cmd += ['--' + key, value]
+        for name, value in docker_args.items():
+            # Passed just the name as boolean flag
+            if isinstance(value, bool) and value:
+                cmd += ['-' + name]
```
Comment (dbczumar) [inline mlflow/projects/__init__.py:540]: I think we need to use a double hyphen for long options (`name` >= 2 chars) and a single hyphen for short options (`name` = 1 char)

**Comment by Zethson** [comment]
@dbczumar `-A t and -A tty` now both work.

File: mlflow/cli.py [hunk 6]
```diff
@@ -52,8 +52,9 @@ def cli():
                    "are not in the list of parameters for an entry point will be passed to the "
                    "corresponding entry point as command-line arguments in the form `--name value`")
 @click.option("--docker-args", "-A", metavar="NAME=VALUE", multiple=True,
-              help="A `docker run` flag or argument, of the form -A name=value. Where `name` "
-              "will then be propagated as `docker run --name value`.")
+              help="A `docker run` argument or flag, of the form -A name=value (e.g. -A gpus=all) "
+                   "or -A name (e.g. -A t). The argument will then be passed as "
+                   "`docker run --name value` or `docker run -name` respectively. ")
```
Comment (dbczumar) [inline mlflow/cli.py:57]: Nit:
```suggestion
                   "`docker run --name value` or `docker run --name` respectively. ")
```

File: mlflow/projects/__init__.py [hunk 7]
```diff
@@ -534,8 +534,16 @@ def _get_docker_command(image, active_run, docker_args=None, volumes=None, user_
     cmd = [docker_path, "run", "--rm"]
 
     if docker_args:
-        for key, value in docker_args.items():
-            cmd += ['--' + key, value]
+        for name, value in docker_args.items():
+            # Passed just the name as boolean flag
+            if isinstance(value, bool) and value:
+                if len(name) == 1:
+                    cmd += ['-' + name]
```
Comment (dbczumar) [inline mlflow/projects/__init__.py:541]: We should do this length check when `value` is not a bool as well. For example, docker's `-u/--user` argument requires a string value; the current implementation would produce `--u value` as opposed to `-u value`, causing an error.

File: mlflow/cli.py [hunk 8]
```diff
@@ -136,16 +137,22 @@ def run(uri, entry_point, version, param_list, docker_args, experiment_name, exp
...
+def _user_args_to_dict(arguments, argument_type='P'):
     user_dict = {}
-    for s in user_list:
-        index = s.find("=")
-        if index == -1:
+    for arg in arguments:
+        split = arg.split('=')
+        # Docker arguments such as `t` don't require a value
+        # -> set them to True if specified
+        if len(split) == 1:
```
Comment (dbczumar) [inline mlflow/cli.py:146]: can we re-introduce the check against `argument_type` so that valueless inputs are only acceptable for `argument_type=A`? We don't currently support valueless inputs for project parameters - if we wanted to do so, it would be best to create a separate PR that also updates the documentation for project parameters.

**Review by dbczumar — CHANGES_REQUESTED** [review]

**Comment by dbczumar** [comment]
@Zethson This is looking great - really close. Excited to get this merged once comments are addressed. Thanks for your contribution!

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @Zethson !
